hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -1725,6 +1729,173 @@ public void delete(final Path path, final boolean recursive,
+          String.format("Path %s is child-blobs", path.toUri().getPath()));
+      if (!recursive) {
+        LOG.error(
+            String.format("Non-recursive delete of non-empty directory %s",
+                path.toUri().getPath()));
+        throw new IOException(
+            "Non-recursive delete of non-empty directory");
+      }
+      ListBlobQueue queue = new ListBlobQueue(blobList.getBlobPropertyList(),
+          getAbfsConfiguration().getProducerQueueMaxSize(),
+          getAbfsConfiguration().getBlobDirDeleteMaxThread());
+      if (blobList.getNextMarker() != null) {
+        new ListBlobProducer(listSrc, client, queue,
+            blobList.getNextMarker(), tracingContext);
+      } else {
+        queue.complete();
+      }
+      ListBlobConsumer consumer = new ListBlobConsumer(queue);
+      deleteCheckOnParentBlob(path, tracingContext);
+      deleteOnConsumedBlobs(path, consumer, tracingContext);
+      return;
+    }
+    LOG.debug(String.format("Path %s doesn't have child-blobs",
+        path.toUri().getPath()));
+    if (!path.isRoot()) {
+      try {
+        getBlobProperty(path, tracingContext);
+      } catch (AbfsRestOperationException ex) {
+        if (ex.getStatusCode() == HTTP_NOT_FOUND) {
+          LOG.error(
+              String.format("Path %s doesn't exist", path.toUri().getPath()));
+          throw new AbfsRestOperationException(
+              ex.getStatusCode(),
+              AzureServiceErrorCode.PATH_NOT_FOUND.getErrorCode(),
+              ex.getErrorMessage(), ex);
+        }
+      }
+      LOG.debug(
+          String.format("Path %s is not a directory", path.toUri().getPath()));
+      deleteCheckOnParentBlob(path, tracingContext);
+      LOG.debug(String.format("Deleting path %s", path.toUri().getPath()));
+      client.deleteBlobPath(path, null, tracingContext);
+      LOG.debug(String.format("Path deleted %s", path.toUri().getPath()));
+    }
+  }
+
+  /**
+   * If parent blob is implicit directory and in case the blob deleted was the
+   * only blob in the directory, it will render path as non-existing. To prevent
+   * that happening, we check if the parent directory is implicit and if yes, we
+   * create marker-based directory.
+   *
+   * @param path path getting deleted
+   * @param tracingContext tracingContext to trace the API flow
+   * @throws IOException
+   */
+  private void deleteCheckOnParentBlob(final Path path,
+      final TracingContext tracingContext) throws IOException {
+    if (path.isRoot()) {
+      return;
+    }
+    Path parentPath = path.getParent();
+    if (parentPath.isRoot()) {
+      return;
+    }
+    LOG.debug(String.format("Checking parent of Path %s : %s exists",
+        path.toUri().getPath(), parentPath.toUri().getPath()));
+    try {
+      getBlobProperty(parentPath, tracingContext);
+      LOG.debug(
+          String.format("Parent of Path %s : %s exists", path.toUri().getPath(),
+              parentPath.toUri().getPath()));
```

Comment:
Shouldn't we have a return statement here ?

Reply:
explicit return would have no role here. Since, there is nothing added after catch block.

Reply:
Bracket confusion happened, makes sense